Receipts

bsonntag opened the pull request:
When an Expense is created it must be linked to a Receipt (except when it's created by a PeriodicExpense).

bsonntag committed:
Creates receipts.

A receipt belongs to an expense and it has a picture.
'carrierwave' gem is used to upload pictures.

bsonntag committed:
Creates receipts.

A receipt belongs to an expense and it has a picture.
'carrierwave' gem is used to upload pictures.

bsonntag committed:
Removed comments from ReceiptPictureUploader.

bsonntag committed:
Addes file uploader to expense form.

bsonntag committed:
Updates create expense integration test.

Also removes ExpenseController tests, as the integration tests make them redundant.

bsonntag committed:
Updates ExpensesController.

ExpensesController now creates a Receipt with the Expense.

bsonntag committed:
Removes gif test image.

bsonntag commented on app/controllers/expenses_controller.rb:
```diff
@@ -10,19 +10,19 @@ def new
-      flash.now[:error] = 'Some fields were left blank'
-      render :new
+      flash[:errors] = 'Some fields were left blank'
+      redirect_to new_expense_path
```
@zamith What do you have to say about this method? I don't really like how it is, because it loses custom errors in the interface...

zamith replied:
Why did you decide to redirect instead of render?

bsonntag replied:
I had to either redirect or do `@expense = Expense.new` before render.

Should CreateExpensesWithReceipt class have a method to retrieve the created expense after calling create? This way I would be able to send the errors to the view.

zamith replied:
Yes, I think that would be better than the redirect in which you loose the errors.

zamith commented on app/models/create_expenses_with_receipt.rb:
```diff
@@ -0,0 +1,29 @@
+    @params = expense_params
+  end
+
+  def create_expense_and_receipt
```
I'm not a big fan of this name, because it is a bit redundant when compared to the class name. Instead, I would just call it `create`.

Or, if you want to make look more like a pattern thing, and make it clearer that it will only have one public method, it can be called `call`, `run`or `process`.

What do you think?

bsonntag replied:
You're right, it's a little redundant. I'll change it to `create`.

bsonntag committed:
Renames CreateExpensesWithReceipt method.

It's now just called create.

bsonntag committed:
Adds tests for CreateExpensesWithReceipt class.

bsonntag commented on app/models/create_expenses_with_receipt.rb:
```diff
@@ -0,0 +1,29 @@
+      create_receipt
+    end
+    @expense
+  end
```
@zamith Should `create` return true/false (depending on whether it was successful or not)? If so, how could I do that here?

zamith replied:
Yes, it should. What happens when a transaction fails? Does it blow up, or it just rolls back and fails silently?

You could assume that if there is a persisted expense then it was successful. In that case, you just return `@expense.persisted?`.

bsonntag committed:
Adds more tests for CreateExpensesWithReceipt.

bsonntag committed:
Refactors CreateExpensesWithReceipt.

Adds an attribute reader for expense and changes create to return true
or false.

bsonntag committed:
Refactors ExpenseController.

ExpenseController#create now renders on failure.

bsonntag commented on app/models/create_expenses_with_receipt.rb:
```diff
@@ -0,0 +1,31 @@
+      create_receipt
+    end
+    @expense.persisted?
+  end
```
@zamith: This fails when some attribute other than picture is missing. For example, if user is missing: "Failure/Error: expect(creator.create).to be false
ActiveRecord::RecordInvalid: Validation failed: User can't be blank"

zamith replied:
What raises that error? You'll probably have to catch that error and return false, with the proper errors set up.

bsonntag replied:
It's being raised by line 25 (`@expense = Expense.create!(@params.except(:picture))`) in the `create_expense` call inside the transaction block.

bsonntag committed:
Refactors CreatesExpensesWithReceipt.

Expenses and receipts are now validated before saving.
Errors from invalid receipts are merged to expense errors.

bsonntag committed:
Adds 'validates_associated :receipt' to Expense.

This fixes the missing "can't be blank" error message in the picture
input of the expense form.

bsonntag commented:
@zamith Is it good now?

zamith commented on app/models/create_expenses_with_receipt.rb:
```diff
@@ -0,0 +1,35 @@
+
+  def create
+    setup
+    unless @expense.valid?
```
What if the expense is valid, but the receipt is not?

zamith replied:
Oh, it's in the model...

bsonntag replied:
Is this bad? Should this check if both are valid?

zamith replied:
No, not really, I just wasn't expecting it.

zamith commented on app/models/create_expenses_with_receipt.rb:
```diff
@@ -0,0 +1,35 @@
+      merge_errors
+      return false
+    end
+    do_saves
```
What about `save`? It would make extra sense because it returns a boolean, such as the AR `save`.

bsonntag committed:
Renames CreateExpensesWithReceipt#do_saves to save

zamith commented:
:+1:
